src/Castle.Core/DynamicProxy/ProxyGenerationOptions.cs

```diff
@@ -204,7 +208,53 @@ public override int GetHashCode()
 			result = 29*result + (Selector != null ? 1 : 0);
 			result = 29*result + MixinData.GetHashCode();
 			result = 29*result + (BaseTypeForInterfaceProxy != null ? BaseTypeForInterfaceProxy.GetHashCode() : 0);
+		    result = 29*result + GetContentsHashCode(AdditionalAttributes);
 			return result;
 		}
+
+        private static int GetContentsHashCode<T>(IList<T> list)
+        {
+            var result = 0;
+            for (var i = 0; i < list.Count; i++)
+            {
+                if (list[i] != null)
+                {
+                    // simply XOR since order does not matter
+                    result ^= list[i].GetHashCode();
+                }
+            }
+
+            return result;
+        }
+
+        private static bool ListsEquivalent<T>(IList<T> listA, IList<T> listB)
+        {
+            if (listA.Count != listB.Count)
+            {
+                return false;
+            }
+
+            // order is not important, just make sure that each entry in A is also found in B

```

Comment:
I'm less worried about `GetHashCode` not taking the value into account than the problem with `Equals`. If `GetHashCode` says more things match than we want, that's inconvenient, but not disastrous (although the algorithm could perhaps be improved by adding hash codes of the elements, rather than `xor`ing, if desired).

Sorry, but I don't follow what you propose for altering the `GetHashCode`.

As far as having `Equals` first compare hash codes, that _could_ work, but really only provides benefits if the hash code is very quick to compute and if we expect to get a lot of equality mismatches.